### PR TITLE
design-refresh - tabs and file sidebar

### LINKED ISSUE #20018: design-refresh - update tabs
### Update our tab elements to match the design refresh

Designs: https://www.figma.com/file/NIsN34NH7lPu04olBzddTw/Design-Refresh-Systemization?node-id=954%3A5153

We should ensure these updates are scoped under `theme-redesign`.

### LINKED ISSUE #20372: Design refresh: File sidebar
Sticker sheet:

https://www.figma.com/file/NIsN34NH7lPu04olBzddTw/Design-Refresh-Systemization-source-of-truth?node-id=955%3A664

### PR DESCRIPTION
**Pull request by 5h1rU**
This PR includes tabs and file sidebar changes

close #20372 
close #20018

🎨 [Figma link ](https://www.figma.com/file/NIsN34NH7lPu04olBzddTw/Design-Refresh-Systemization-source-of-truth?node-id=954%3A8291)

https://user-images.githubusercontent.com/989826/116623336-b6b54700-a90b-11eb-9161-7ab95c4a9850.mov

### CODE REVIEW

File: client/branded/src/components/panel/Panel.tsx [hunk 1]
```diff
@@ -237,9 +237,13 @@ export const Panel = React.memo<Props>(props => {
         <Tabs className="panel" index={tabIndex} onChange={handleActiveTab}>
             <div className="tablist-wrapper bg-body d-flex justify-content-between">
                 <TabList>
-                    {items.map(({ label, id }) => (
-                        <Tab key={id}>{label}</Tab>
-                    ))}
+                    <div className="border-bottom d-flex w-100">
+                        {items.map(({ label, id }) => (
+                            <Tab key={id}>
+                                <span className="tablist-label">{label}</span>
```
Comment (pdubroy) [inline client/branded/src/components/panel/Panel.tsx:243]: It feels like we should create our own `Tab` component that knows how to style the label. That would also allow us to avoid introducing a new global class name.

File: client/web/src/repo/RepoRevisionSidebar.tsx [hunk 2]
```diff
@@ -81,7 +85,7 @@ export const RepoRevisionSidebar: React.FunctionComponent<Props> = props => {
                         aria-hidden={true}
                         className="d-flex overflow-auto repo-revision-container__tabpanels explorer"
                     >
-                        <TabPanels className="w-100">
+                        <TabPanels className="w-100 tablist-content-wrapper">
```
Comment (pdubroy) [inline client/web/src/repo/RepoRevisionSidebar.tsx:88]: Similar to `Tab`, we could introduce our own `TabPanels` component too.

File: client/web/src/repo/RevisionsPopover.tsx [hunk 3]
```diff
@@ -224,18 +228,30 @@ export const RevisionsPopover: React.FunctionComponent<Props> = props => {
...
-                            {label}
+                        <Tab className="tablist-wrapper__full-width" key={id} data-test-tab={id}>
+                            <span className="tablist-label">{label}</span>
                         </Tab>
                     ))}
                 </TabList>
+                {isRedesignEnabled && (
+                    <button
+                        type="button"
+                        // onClick={handlePanelClose}
```
Comment (pdubroy) [inline client/web/src/repo/RevisionsPopover.tsx:242]: Remove commented out code.

File: client/web/src/tree/util.tsx [hunk 4]
```diff
@@ -41,11 +41,13 @@ export function scrollIntoView(element: Element, scrollRoot: Element): void {
     scrollRoot.scrollLeft = 0
 }
 
+const isRedesingEnabled: string | null = localStorage.getItem('isRedesignEnabled')
```
Comment (pdubroy) [inline client/web/src/tree/util.tsx:44]:
```suggestion
const isRedesignEnabled: string | null = localStorage.getItem('isRedesignEnabled')
```

File: client/web/src/tree/util.tsx [hunk 5]
```diff
@@ -41,11 +41,13 @@ export function scrollIntoView(element: Element, scrollRoot: Element): void {
...
+
 export const getDomElement = (path: string): Element | null =>
     document.querySelector(`[data-tree-path='${path.replace(/'/g, "\\'")}']`)
 
 export const treePadding = (depth: number, isTree: boolean): React.CSSProperties => ({
-    marginLeft: `${depth * 12 + (isTree ? 0 : 12) + 12}px`,
+    marginLeft: `${depth * 12 + (isTree ? 0 : 12) + (isRedesingEnabled ? 0 : 12)}px`,
```
Comment (pdubroy) [inline client/web/src/tree/util.tsx:50]:
```suggestion
    marginLeft: `${depth * 12 + (isTree ? 0 : 12) + (isRedesignEnabled ? 0 : 12)}px`,
```

File: client/branded/src/components/panel/Panel.tsx (hunk 1 above)
Comment (5h1rU) [reply client/branded/src/components/panel/Panel.tsx:243]: Actually, I think this is a huge advantage when you want to make small modifications to different areas where the component is used. In this case, the tabs design is slightly different comparing with the other two places where those are being used.

Comment (pdubroy) [reply client/branded/src/components/panel/Panel.tsx:243]: Isn't this an example of our "standard" tab? If I'm implementing a new UI feature that has tabs in it, I shouldn't have to use `<span className="tablist-label">` just to achieve the default behavior that nearly all tabs should have.

It's good to support customization, but we should optimize for the common case, not the exceptions. It looks `tablist-label` is required to implement the focus state correctly so I'm wondering why we don't want to have that (nearly) everywhere.

Comment (5h1rU) [reply client/branded/src/components/panel/Panel.tsx:243]: I agree that this repetitive pattern can be part of a separate component. I create [this issue](https://github.com/sourcegraph/sourcegraph/issues/20663) to add Tabs as part of our wildcard library. In the meantime, this PR can handle the look and feel and cover all the cases on design refresh.

File: client/web/src/tree/Tree.scss [hunk 6]
```diff
@@ -33,8 +33,8 @@
 
         &--active {
             width: 100%;
-            background-color: var(--primary) !important;
-            color: $white !important;
+            background-color: var(--primary);
+            color: $white;
```
Comment (vovakulikov) [inline client/web/src/tree/Tree.scss:37]: Shoud we also use here some css variable color instead of sass variable?

File: client/web/src/tree/Directory.tsx [hunk 7]
```diff
@@ -28,66 +29,70 @@ interface TreeChildProps extends TreeLayerProps {
...
+        <tr key={props.entryInfo.path} className={props.className} onClick={props.handleTreeClick}>
+            <td className="tree__cell test-sidebar-file-decorable">
+                <div
+                    className="tree__row-contents tree__row-contents-new"
+                    data-tree-is-directory="true"
+                    data-tree-path={props.entryInfo.path}
+                >
+                    <div className="tree__row-contents-text flex-1 justify-between">
+                        <div className="d-flex">
+                            <a
```
Comment (vovakulikov) [inline client/web/src/tree/Directory.tsx:45]: I know this is not the part of this ticket but I'm curious what is the reason to have two separate links one for icon and one for text? And with turned off tab-index. @5h1rU maybe you know since you've been here.

File: client/web/src/tree/Tree.scss (hunk 6 above)
Comment (5h1rU) [reply client/web/src/tree/Tree.scss:37]: We currently don't have any CSS variable for white value.

This is part of the original component, which will be replaced with the new design. I think is better to keep this value as it is.

File: client/web/src/tree/Directory.tsx (hunk 7 above)
Comment (5h1rU) [reply client/web/src/tree/Directory.tsx:45]: Me neither. This seems like at some point, this was having different behaviors depends on the case. By now, both links are pointing to the same reference.

Since we focus entirely on the look and feel, I don't want to touch this logic for now.

I'll take a note and address this concern when we are working on improving these components.

File: client/web/src/repo/RevisionsPopover.tsx [hunk 8]
```diff
@@ -224,18 +228,30 @@ export const RevisionsPopover: React.FunctionComponent<Props> = props => {
 
     return (
         <Tabs defaultIndex={tabIndex} className="revisions-popover" onChange={handleTabsChange}>
-            <div className="tablist-wrapper w-100 align-items-center">
+            <div className="tablist-wrapper flex-1">
                 <TabList>
                     {TABS.map(({ label, id }) => (
-                        <Tab className="d-flex flex-1 justify-content-around" key={id} data-test-tab={id}>
-                            {label}
+                        <Tab className="tablist-wrapper__full-width" key={id} data-test-tab={id}>
```
Comment (pdubroy) [inline client/web/src/repo/RevisionsPopover.tsx:234]: To follow BEM, I think this should be `tablist-wrapper--full-width` since `full-width` is a modifier, not an element.

File: client/web/src/repo/RevisionsPopover.tsx [hunk 9]
```diff
@@ -224,18 +228,30 @@ export const RevisionsPopover: React.FunctionComponent<Props> = props => {
...
+                            <span className="tablist-label">{label}</span>
```
Comment (pdubroy) [inline client/web/src/repo/RevisionsPopover.tsx:235]: This class name also doesn't seem to follow BEM properly, because the `label` part represents the _element_. It's a bit verbose, but I'd suggest `tablist-wrapper--tab-label` for consistency (since the encloding BEM block is `tablist-wrapper`) and so that we don't need to introduce any unnecessary other class names.

File: client/branded/src/global-styles/tabs.scss [hunk 10]
```diff
@@ -29,6 +25,66 @@
...
+        }
+        &[data-selected] {
+            color: var(--body-color);
+            font-weight: bold;
+            border-bottom: 3px solid var(--brand-secondary);
+        }
+        &:focus-visible {
+            outline: none;
+            box-shadow: none;
+            .tablist-label {
```
Comment (pdubroy) [inline client/branded/src/global-styles/tabs.scss:71]: Our style guide says to avoid descendant and child selectors, though I think this is a case where it makes sense. But, it's preferred to use a child selector if we don't actually need arbitrary-depth descendants.

```suggestion
            > .tablist-label {
```

File: client/branded/src/global-styles/tabs.scss [hunk 11]
```diff
@@ -29,6 +25,66 @@
...
+    }
+
+    .tablist-wrapper {
+        min-height: 2rem;
+        border-bottom: 1px solid var(--border-color);
+        padding-bottom: 0;
+        display: flex;
+        align-items: stretch;
+        justify-content: space-between;
+        &__full-width {
```
Comment (pdubroy) [inline client/branded/src/global-styles/tabs.scss:86]: Why is this needed? Looks redundant with the `display: flex` on line 83.

File: client/web/src/repo/RevisionsPopover.tsx [hunk 12]
```diff
@@ -224,18 +228,30 @@ export const RevisionsPopover: React.FunctionComponent<Props> = props => {
...
                         </Tab>
                     ))}
                 </TabList>
+                {isRedesignEnabled && (
+                    <button
+                        type="button"
+                        // TODO: add onClick funcitonality
```
Comment (pdubroy) [inline client/web/src/repo/RevisionsPopover.tsx:242]: Can you please open a ticket and link it in the TODO here?

```suggestion
                        // TODO: add onClick functionality
```

Comment (pdubroy) [reply client/web/src/repo/RevisionsPopover.tsx:242]: Actually, given that this button is not functional and there's very little code here, it's probably better to remove the code here from this PR, and add it only when you implement the onClick functionality. Overall I'm in favour of working incrementally, and I believe that it's totally ok to have incomplete code that is inactive, but it doesn't seem worth it in this case.

File: client/web/src/tree/util.tsx [hunk 13]
```diff
@@ -44,8 +44,8 @@ export function scrollIntoView(element: Element, scrollRoot: Element): void {
 export const getDomElement = (path: string): Element | null =>
     document.querySelector(`[data-tree-path='${path.replace(/'/g, "\\'")}']`)
 
-export const treePadding = (depth: number, isTree: boolean): React.CSSProperties => ({
-    marginLeft: `${depth * 12 + (isTree ? 0 : 12) + 12}px`,
+export const treePadding = (depth: number, isTree: boolean, isRedesignEnabled?: boolean): React.CSSProperties => ({
```
Comment (pdubroy) [inline client/web/src/tree/util.tsx:47]:
```suggestion
export const treePadding = (depth: number, isTree: boolean, isRedesignEnabled=false): React.CSSProperties => ({
```

There's little actual benefit here, but generally I think it's good practice to set a default value of false in cases like this, because it results in a tighter type — inside the function, only `true` and `false` are valid values (and not `undefined`).

**Review by pdubroy — APPROVED** [review]
lgtm with some minor suggestions / questions

File: client/branded/src/global-styles/tabs.scss [hunk 14]
```diff
@@ -29,6 +25,53 @@
 }
 
 .tablist-wrapper {
-    background: var(--body-bg);
     border-bottom: 3px solid var(--border-color);
+    align-items: center;
+    &__full-width {
```
Comment (pdubroy) [inline client/branded/src/global-styles/tabs.scss:30]: @5h1rU is this used anywhere?
